ibis/backends/duckdb/__init__.py

```diff
@@ -534,6 +544,25 @@ def attach_sqlite(
                 sa.text(f"CALL sqlite_attach('{str(path)}', overwrite={overwrite})")
             )
 
+    def _register_in_memory_tables(self, expr: ir.Expr) -> None:
+        from ibis.expr.analysis import find_physical_tables
+
+        # Warn for any tables depending on RecordBatchReaders that have already
+        # started being consumed.
+        for t in find_physical_tables(expr.op()):
```

Comment:
This hook needs to be run before every query, same as `_register_in_memory_tables`. I piggy-backed on the same method here for now, but perhaps we should refactor to a general `_pre_execute(expr: Expr) -> None` hook, the default implementation of which just calls `_register_in_memory_tables`, just so we don't add stuff that isn't memtable related here.

Reply:
Yeah, +1 on refactoring this to something slightly more generic.